File: android/app/src/main/java/com/raban/etabli/focus/engine/PomodoroEngine.kt
// Copyright 2026 Raban Heller
// SPDX-License-Identifier: Apache-2.0

package com.raban.etabli.focus.engine

import android.content.Context
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.booleanPreferencesKey
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.intPreferencesKey
import androidx.datastore.preferences.core.longPreferencesKey
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import com.raban.etabli.focus.data.AppDatabase
import com.raban.etabli.focus.data.SessionEntity
import com.raban.etabli.focus.notifications.NotificationScheduler
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import java.util.UUID

// Android twin of the iOS PomodoroEngine (PomodoroEngine.swift).
//
// Identical contract:
//   - active timer state lives in DataStore (not Room) so it survives
//     force-quit / background / reboot AND can be read instantly on launch
//     before the @Database container opens
//   - `endEpochMillis` is the source of truth for remaining time while
//     running (NOT a wall-clock tick) — guarantees correctness across
//     backgrounding
//   - `pausedRemainingMillis` carries the budget while paused
//   - tick() detects the "ran out while we were backgrounded" case and
//     finalises the SessionEntity row with completed=true
//   - notifications scheduled via AlarmManager so they fire at endDate
//     regardless of process state.

enum class TimerPhase { Idle, Running, Paused }
enum class TimerKind  { Focus, ShortBreak, LongBreak }

data class TimerState(
    val phase: TimerPhase = TimerPhase.Idle,
    val kind: TimerKind = TimerKind.Focus,
    val categoryID: String? = null,
    val sessionID: String? = null,
    val endEpochMillis: Long? = null,
    val pausedRemainingMillis: Long? = null,
    val totalSeconds: Long = 0,
    val blocksCompletedInSet: Int = 0,
)

// MARK: - DataStore-backed persistence

private val Context.timerStore by preferencesDataStore(name = "timer_state")

private object Keys {
    val phase            = stringPreferencesKey("phase")
    val kind             = stringPreferencesKey("kind")
    val categoryID       = stringPreferencesKey("categoryID")
    val sessionID        = stringPreferencesKey("sessionID")
    val endEpoch         = longPreferencesKey("endEpoch")
    val pausedRemaining  = longPreferencesKey("pausedRemaining")
    val totalSeconds     = longPreferencesKey("totalSeconds")
    val blocksInSet      = intPreferencesKey("blocksInSet")
    val touched          = booleanPreferencesKey("touched")
}

class PomodoroEngine(
    private val context: Context,
    private val db: AppDatabase,
    private val scope: CoroutineScope,
) {
    val state: StateFlow<TimerState> = context.timerStore.data
        .map { prefs -> readState(prefs) }
        .stateIn(scope, started = kotlinx.coroutines.flow.SharingStarted.Eagerly, TimerState())

    private val scheduler = NotificationScheduler(context)

    /// Start a fresh focus/break of `minutes`. Persists end-epoch and
    /// schedules a completion alarm.
    fun start(categoryID: String, kind: TimerKind, minutes: Int) {
        val seconds = minutes * 60L
        val now = System.currentTimeMillis()
        val end = now + seconds * 1000
        val sessionID = UUID.randomUUID().toString()
        scope.launch {
            context.timerStore.edit { p ->
                p[Keys.phase] = TimerPhase.Running.name
                p[Keys.kind] = kind.name
                p[Keys.categoryID] = categoryID
                p[Keys.sessionID] = sessionID
                p[Keys.endEpoch] = end
                p.remove(Keys.pausedRemaining)
                p[Keys.totalSeconds] = seconds
                p[Keys.touched] = true
            }
            scheduler.scheduleCompletion(at = end, kind = kind)
        }
    }

    fun pause() {
        scope.launch {
            val s = state.value
            if (s.phase != TimerPhase.Running) return@launch
            val remaining = ((s.endEpochMillis ?: 0) - System.currentTimeMillis()).coerceAtLeast(0)
            context.timerStore.edit { p ->
                p[Keys.phase] = TimerPhase.Paused.name
                p.remove(Keys.endEpoch)
                p[Keys.pausedRemaining] = remaining
            }
            scheduler.cancel()
        }
    }

    fun resume() {
        scope.launch {
            val s = state.value
            val rem = s.pausedRemainingMillis ?: return@launch
            val newEnd = System.currentTimeMillis() + rem
            context.timerStore.edit { p ->
                p[Keys.phase] = TimerPhase.Running.name
                p[Keys.endEpoch] = newEnd
                p.remove(Keys.pausedRemaining)
            }
            scheduler.scheduleCompletion(at = newEnd, kind = s.kind)
        }
    }

    /// Cancel without completing — writes a SessionEntity with completed=false
    /// reflecting actual elapsed time.
    fun cancel() {
        scope.launch {
            val s = state.value
            if (s.sessionID != null) writeSession(s, completed = false)
            context.timerStore.edit { it.clear() }
            scheduler.cancel()
        }
    }

    /// Detects "ran out while we were backgrounded" and finalises the session.
    /// Called from a Compose `LaunchedEffect` once per second.
    fun tick() {
        scope.launch {
            val s = state.value
            if (s.phase != TimerPhase.Running) return@launch
            val end = s.endEpochMillis ?: return@launch
            if (System.currentTimeMillis() >= end) {
                if (s.kind == TimerKind.Focus) {
                    writeSession(s, completed = true)
                }
                context.timerStore.edit { p ->
                    p[Keys.phase] = TimerPhase.Idle.name
                    p.remove(Keys.endEpoch)
                    p.remove(Keys.sessionID)
                    if (s.kind == TimerKind.Focus) {
                        p[Keys.blocksInSet] = (p[Keys.blocksInSet] ?: 0) + 1
                    }
                }
            }
        }
    }

    private suspend fun writeSession(s: TimerState, completed: Boolean) {
        val cid = s.categoryID ?: return
        val sid = s.sessionID ?: return
        val now = System.currentTimeMillis()
        val startedAt = (s.endEpochMillis ?: now) - s.totalSeconds * 1000
        db.sessionDao().insert(
            SessionEntity(
                id = sid,
                categoryID = cid,
                startedAtEpoch = startedAt,
                plannedEndEpoch = s.endEpochMillis ?: now,
                actualEndEpoch = now,
                completed = completed,
                kind = if (s.kind == TimerKind.Focus) "focus" else "break",
            )
        )
    }

    private fun readState(p: Preferences): TimerState = TimerState(
        phase                  = TimerPhase.valueOf(p[Keys.phase] ?: TimerPhase.Idle.name),
        kind                   = TimerKind.valueOf(p[Keys.kind]   ?: TimerKind.Focus.name),
        categoryID             = p[Keys.categoryID],
        sessionID              = p[Keys.sessionID],
        endEpochMillis         = p[Keys.endEpoch],
        pausedRemainingMillis  = p[Keys.pausedRemaining],
        totalSeconds           = p[Keys.totalSeconds] ?: 0,
        blocksCompletedInSet   = p[Keys.blocksInSet] ?: 0,
    )

    companion object {
        @Volatile private var instance: PomodoroEngine? = null
        fun get(context: Context, db: AppDatabase, scope: CoroutineScope): PomodoroEngine {
            return instance ?: synchronized(this) {
                instance ?: PomodoroEngine(context.applicationContext, db, scope).also { instance = it }
            }
        }
    }
}

/// Helper for the UI: seconds remaining, computed from the state snapshot.
fun TimerState.remainingSeconds(now: Long = System.currentTimeMillis()): Long {
    return when (phase) {
        TimerPhase.Running -> ((endEpochMillis ?: 0) - now).coerceAtLeast(0) / 1000
        TimerPhase.Paused  -> (pausedRemainingMillis ?: 0) / 1000
        TimerPhase.Idle    -> 0
    }
}
/// 0..1 progress for the ring.
fun TimerState.progress(now: Long = System.currentTimeMillis()): Float {
    if (totalSeconds == 0L) return 0f
    val remaining = remainingSeconds(now)
    return (1f - remaining.toFloat() / totalSeconds.toFloat()).coerceIn(0f, 1f)
}

File: ios/App/PomodoroEngine.swift
// Copyright 2026 Raban Heller
// SPDX-License-Identifier: Apache-2.0

import Foundation
import Observation
import SwiftUI
import SwiftData
import UserNotifications

// MARK: - Persisted timer state
//
// Active-timer state lives in UserDefaults as JSON. Living outside
// SwiftData on purpose: the timer must survive force-quit / background /
// device restart and be reconstructable BEFORE the @Model container
// finishes opening (otherwise an "almost done" session could lose seconds).
public struct TimerState: Codable, Equatable {
    public enum Kind: String, Codable { case focus, breakShort, breakLong }
    public enum Phase: String, Codable { case running, paused, idle }
    public var phase: Phase
    public var kind: Kind
    public var categoryID: UUID?
    public var sessionID: UUID?
    /// `endDate` is the source of truth for time-remaining while running.
    /// `pausedRemainingSeconds` carries over the budget while paused.
    public var endDate: Date?
    public var pausedRemainingSeconds: Double?
    public var totalSeconds: Double
    public var blocksCompletedInSet: Int   // for "long break after N"
    public static let idle = TimerState(phase: .idle, kind: .focus,
                                        categoryID: nil, sessionID: nil,
                                        endDate: nil, pausedRemainingSeconds: nil,
                                        totalSeconds: 0, blocksCompletedInSet: 0)
}

// MARK: - Engine

@MainActor
@Observable
public final class PomodoroEngine {
    public private(set) var state: TimerState
    public var blocksUntilLongBreak: Int = 4
    public var longBreakMinutes: Int = 20

    private let defaultsKey = "scholarfocus.timer.state"
    private let notificationID = "scholarfocus.completion"

    public init() {
        if let data = UserDefaults.standard.data(forKey: defaultsKey),
           let restored = try? JSONDecoder().decode(TimerState.self, from: data) {
            self.state = restored
        } else {
            self.state = .idle
        }
    }

    // MARK: - Derived view

    public var remainingSeconds: Double {
        switch state.phase {
        case .running:
            guard let end = state.endDate else { return 0 }
            return max(0, end.timeIntervalSinceNow)
        case .paused:
            return max(0, state.pausedRemainingSeconds ?? 0)
        case .idle:
            return 0
        }
    }

    /// 0…1 progress for the ring.
    public var progress: Double {
        guard state.totalSeconds > 0 else { return 0 }
        return 1 - (remainingSeconds / state.totalSeconds)
    }

    // MARK: - Control

    public func start(categoryID: UUID, kind: TimerState.Kind, minutes: Int) {
        let seconds = Double(minutes) * 60
        let now = Date()
        var s = state
        s.phase = .running
        s.kind = kind
        s.categoryID = categoryID
        s.sessionID = UUID()
        s.endDate = now.addingTimeInterval(seconds)
        s.pausedRemainingSeconds = nil
        s.totalSeconds = seconds
        state = s
        persist()
        scheduleNotification(at: s.endDate!, kind: kind)
    }

    public func pause() {
        guard state.phase == .running else { return }
        let remaining = remainingSeconds
        var s = state
        s.phase = .paused
        s.endDate = nil
        s.pausedRemainingSeconds = remaining
        state = s
        persist()
        cancelNotification()
    }

    public func resume() {
        guard state.phase == .paused, let remaining = state.pausedRemainingSeconds else { return }
        var s = state
        s.phase = .running
        s.endDate = Date().addingTimeInterval(remaining)
        s.pausedRemainingSeconds = nil
        state = s
        persist()
        scheduleNotification(at: s.endDate!, kind: s.kind)
    }

    /// Cancel without completing — records a Session with `completed = false`.
    public func cancel(in context: ModelContext) {
        if state.sessionID != nil {
            writeSession(into: context, completed: false)
        }
        state = .idle
        persist()
        cancelNotification()
    }

    /// Detects "ran out while we were backgrounded" and finalises the session.
    /// Call from the UI's TimelineView once per second.
    public func tick(in context: ModelContext) {
        guard state.phase == .running, let end = state.endDate else { return }
        if Date() >= end {
            if state.kind == .focus {
                writeSession(into: context, completed: true)
                state.blocksCompletedInSet += 1
            }
            state.phase = .idle
            state.endDate = nil
            state.sessionID = nil
            persist()
        }
    }

    // MARK: - Session writes

    private func writeSession(into context: ModelContext, completed: Bool) {
        guard let cid = state.categoryID, let sid = state.sessionID else { return }
        let actualEnd = Date()
        let started = (state.endDate ?? actualEnd).addingTimeInterval(-state.totalSeconds)
        let session = Session(
            id: sid,
            categoryID: cid,
            startedAt: started,
            plannedEnd: state.endDate ?? actualEnd,
            actualEnd: actualEnd,
            completed: completed,
            kind: state.kind == .focus ? "focus" : "break"
        )
        context.insert(session)
        try? context.save()
    }

    // MARK: - Persistence

    private func persist() {
        if let data = try? JSONEncoder().encode(state) {
            UserDefaults.standard.set(data, forKey: defaultsKey)
        }
    }

    // MARK: - Notifications

    /// Asks for permission. Safe to call repeatedly.
    public static func requestAuthorization() async {
        let center = UNUserNotificationCenter.current()
        _ = try? await center.requestAuthorization(options: [.alert, .sound, .badge])
    }

    private func scheduleNotification(at date: Date, kind: TimerState.Kind) {
        let center = UNUserNotificationCenter.current()
        center.removePendingNotificationRequests(withIdentifiers: [notificationID])
        let content = UNMutableNotificationContent()
        content.title = kind == .focus ? "Focus block complete" : "Break complete"
        content.body  = kind == .focus
            ? "Suggest a short break — or start the next block."
            : "Time to get back to it when you're ready."
        content.sound = .default
        let interval = max(1, date.timeIntervalSinceNow)
        let trigger = UNTimeIntervalNotificationTrigger(timeInterval: interval, repeats: false)
        let request = UNNotificationRequest(identifier: notificationID, content: content, trigger: trigger)
        center.add(request) { _ in }
    }

    private func cancelNotification() {
        UNUserNotificationCenter.current().removePendingNotificationRequests(
            withIdentifiers: [notificationID]
        )
    }
}
